src/text.c

```diff
@@ -0,0 +1,358 @@
+ *   some=stuff in the sub-namespace
+ *
+ *   /subns/deeper
+ *   this=is in /subns/deeper
+ *
+*/
+
+static int cmp_ns(const void *a, const void *b) {
+	const SdbNs *nsa = a;
+	const SdbNs *cia = b;
+	return strcmp (nsa->name, cia->name);
+}
+
+static bool text_fsave(Sdb *s, FILE *f, bool sort, SdbList *path) {
+// n = position we are currently looking at
+// p = position until we have already written everything
+// macro to flush a block of text that doesn't have to be escaped
+#define FLUSH do { if (p != n) { fwrite (p, 1, n - p, f); p = n; } } while (0)
+// macro to flush and skip a char
+#define SKIP do { FLUSH; p++; } while (0)
+	// path
+	fwrite ("/", 1, 1, f); // always print a /, even if path is empty
+	SdbListIter *it;
+	const char *path_token;
+	bool first = true;
+	ls_foreach (path, it, path_token) {
+		if (first) {
+			first = false;
+		} else {
+			fwrite ("/", 1, 1, f);
+		}
+		// write and escape the path
+		const char *p = path_token;
+		const char *n = p;
+		while (*n) {
+			switch (*n) {
+			case '\\':
+				SKIP;
+				fwrite ("\\\\", 1, 2, f);
+				break;
+			case '/':
+				SKIP;
+				fwrite ("\\/", 1, 2, f);
+				break;
+			case '\n':
+				SKIP;
+				fwrite ("\\n", 1, 2, f);
+				break;
+			}
+			n++;
+		}
+		FLUSH;
+	}
+	fwrite ("\n", 1, 1, f);
+
+	// k=v entries
+	SdbList *l = sdb_foreach_list (s, sort);
+	SdbKv *kv;
+	ls_foreach (l, it, kv) {
+		const char *k = sdbkv_key (kv);
+		const char *v = sdbkv_value (kv);
+		// escape leading '/'
+		if (*k == '/') {
+			fwrite ("\\", 1, 1, f);
+		}
+
+		// write and escape key
+		const char *p = k;
+		const char *n = p;
+		while (*n) {
+			switch (*n) {
+			case '\\':
+				SKIP;
+				fwrite ("\\\\", 1, 2, f);
+				break;
+			case '=':
+				SKIP;
+				fwrite ("\\=", 1, 2, f);
+				break;
+			case '\n':
+				SKIP;
+				fwrite ("\\n", 1, 2, f);
+				break;
+			}
+			n++;
+		}
+		FLUSH;
+
+		fwrite ("=", 1, 1, f);
+
+		// write and escape value
+		p = v;
+		n = p;
+		while (*n) {
+			switch (*n) {
+			case '\\':
+				SKIP;
+				fwrite ("\\\\", 1, 2, f);
+				break;
+			case '\n':
+				SKIP;
+				fwrite ("\\n", 1, 2, f);
+				break;
+			}
+			n++;
+		}
+		FLUSH;
+		fwrite ("\n", 1, 1, f);
+	}
+	ls_free (l);
+
+	// sub-namespaces
+	l = s->ns;
+	if (sort) {
+		l = ls_clone (l);
+		ls_sort (l, cmp_ns);
+	}
+	SdbNs *ns;
+	ls_foreach (l, it, ns) {
+		fwrite ("\n", 1, 1, f);
+		ls_push (path, ns->name);
+		text_fsave (ns->sdb, f, sort, path);
+		ls_pop (path);
+	}
+	if (l != s->ns) {
+		ls_free (l);
+	}
+
+#undef FLUSH
+	return true;
+}
+
+SDB_API bool sdb_text_fsave(Sdb *s, FILE *f, bool sort) {
+	SdbList *path = ls_new ();
+	if (!path) {
+		return false;
+	}
+	bool r = text_fsave (s, f, sort, path);
+	ls_free (path);
+	return r;
+}
+
+SDB_API bool sdb_text_save(Sdb *s, const char *file, bool sort) {
+	FILE *f = fopen (file, "w");
+	if (!f) {
+		return false;
+	}
+	bool r = sdb_text_fsave (s, f, sort);
+	fclose (f);
+	return r;
+}
+
+SDB_API bool sdb_text_fload(Sdb *s, FILE *f) {
```

Comment:
this function is too long. can be split into small functions to make the main parsing loop easier to read